bytecomp/bytelink.ml

```diff
@@ -590,6 +608,49 @@ let output_cds_file outfile =
        Bytesections.write_toc_and_trailer toc_writer;
     )
 
+(* [c_string_literal_of_string s] returns the C literal string representation of
+   [s], suitable for embedding in a C source file with type [char_os *]. The
+   result includes the quote markers. *)
+let c_string_literal_of_string s =
+  let b = Buffer.create (String.length s * 2) in
+  let utf16le = Bytes.create 4 in
+  let escape u =
+    match Uchar.to_int u with
+      (* Characters with C escape sequences *)
+    | 000 (* '\0' *) -> Buffer.add_string b "\\0"
+    | 009 (* '\t' *) -> Buffer.add_string b "\\t"
+    | 010 (* '\n' *) -> Buffer.add_string b "\\n"
+    | 013 (* '\r' *) -> Buffer.add_string b "\\r"
+    | 034 (* '\"' *) -> Buffer.add_string b "\\\""
+    | 092 (* '\\' *) -> Buffer.add_string b "\\\\"
```

Comment:
You may also want to escape `?` to avoid trigraph problems 😖

Reply:
I agree with all of the above, and trigraphs have been removed from C23, which GCC15 defaults to.
- https://en.cppreference.com/w/c/language/operator_alternative.html#Trigraphs_.28removed_in_C23.29
- https://gcc.gnu.org/onlinedocs/cpp/Initial-processing.html